lib/mix/ecto.ex

```diff
@@ -118,6 +118,87 @@ defmodule Mix.Ecto do
   end
 
   @doc """
+  Gets the storage path from the repository.
+  """
+  @spec storage_path(Ecto.Repo.t) :: String.t
+  def storage_path(repo) do
+    repo_priv(repo)
+  end
+
+  @doc """
+  Add me
+  """
+  @spec storage_filename(Ecto.Repo.t) :: String.t
+  def storage_filename(repo) do
+    database_info = repo.__adapter__.database_info
+    "#{inspect(repo)}-#{database_info.type}-#{database_info.version}.storage"
+  end
+
+  @doc """
+  Dumps the storage.
+  """
+  @spec storage_dump(Ecto.Repo.t) :: no_return
+  def storage_dump(repo) do
+    adapter = repo.__adapter__
+    dump = adapter.storage_dump(repo.config)
+  end
+
+  @doc """
+  Loads the storage. Will raise an exception if there is a database
+  type/version mismatch.
+  """
+  @spec storage_load(Ecto.Repo.t, String) :: no_return
+  def storage_load(repo, filename) do
+    adapter = repo.__adapter__
+
+    if database_matches?(repo, filename) do
+      adapter.storage_load(repo.config, filename)
+    else
+      raise "storage Info Mismatch"
+    end
+  end
+
+  @doc """
+  Determines if the database defined in the Repo matches the
+  meta data defined in the storage
+  """
+  @spec database_matches?(Ecto.Repo.t, String) :: boolean
+  def database_matches?(repo, filename) do
+    adapter = repo.__adapter__
+    storage_info = extract_storage_info(adapter, filename)
+    database_info = adapter.database_info
+
+    database_match?(storage_info, database_info) && version_match?(repo, storage_info, database_info)
+  end
+  defp database_match?(storage_info, database_info) do
+    storage_info[:type] == database_info[:type]
+  end
+  defp version_match?(repo, storage_info, database_info) do
```

Comment:
:-1: for this dance. Here is another approach: we will always try to load some particular version. If the version is not supported, you can pass an explicit file to the schema.load/dump, this way we don't need to guess and it is more explicit.